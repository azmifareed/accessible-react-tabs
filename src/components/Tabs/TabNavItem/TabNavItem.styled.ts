import styled, { css } from 'styled-components';

interface IStyledTab {
  isSelected?: boolean;
}

export const StyledTab = styled.button<IStyledTab>`
  font-size: 1.0625rem;
  font-family: ${p => p.theme.typography.fontFamily};
  font-weight:  ${p => p.theme.typography.fontWeight.regular};
  padding: ${p => `${p.theme.spacing.small} ${p.theme.spacing.large}`};
  color: ${p => p.theme.colors.textDefault};
  cursor: pointer;
  border: 0;
  background: transparent;
  outline: 0;
  position: relative;
  &::after {
    ${p => p.isSelected ? css`
      content: '';
      border-bottom: ${p => `4px solid ${p.theme.colors.secondary}`};
      width: 100%;
      position: absolute;
      left: 0;
      bottom: 0;
    ` : null}
  }
`
