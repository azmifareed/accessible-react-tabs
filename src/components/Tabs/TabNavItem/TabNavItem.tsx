import React from 'react';
import { StyledTab } from './TabNavItem.styled';

export interface ITabNavItemProps {
  children: string;
  tabIndex?: number;
  isSelected?: boolean;
  id: string;
  tabRef: (instance: HTMLButtonElement) => void;
  index: number;
  ariaControls: string;
  onSelect: (event: MouseEvent, index: number, id: string) => void;
  onKeyDown: (event: any) => void;
}

export function TabNavItem({ children, tabIndex, isSelected, id, tabRef, index, ariaControls, onSelect, onKeyDown }: ITabNavItemProps) {
  return (
    <StyledTab
      ref={tabRef}
      role="tab"
      aria-selected={!!isSelected}
      aria-controls={ariaControls}
      id={id}
      tabIndex={tabIndex}
      onClick={(event: any) => onSelect(event, index, id)}
      onKeyDown={onKeyDown}
      isSelected={isSelected}
    >{children}</StyledTab>
  );
}
