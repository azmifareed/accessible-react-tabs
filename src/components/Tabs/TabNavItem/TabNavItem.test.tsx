import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import { render } from '../../../utils/test-utils';
import { TabNavItem, ITabNavItemProps } from './TabNavItem';

const id = 'test';
const tabRef = {
  current: {
    focus: jest.fn()
  }
}
const onSelect = jest.fn();
const onKeyDown = jest.fn();

const initProps = {
  id,
  tabRef: () => tabRef.current,
  isSelected: false,
  ariaControls: `${id}-tab`,
  index: 0,
  tabIndex: -1,
  onSelect,
  onKeyDown,
  children: 'tab text'
}

function renderTabNavItem(props: ITabNavItemProps) {
  return render(<TabNavItem {...props} />)
}

test('renders TabNavItem text', () => {
  renderTabNavItem(initProps);
  const linkElement = screen.getByText(/tab text/i);
  expect(linkElement).toBeInTheDocument();
});

test('renders TabNavItem correctly', () => {
  const component = renderTabNavItem(initProps);
  expect(component.baseElement).toMatchSnapshot();
});

test('renders TabNavItem correctly when selected', () => {
  const component = renderTabNavItem({...initProps, isSelected: true});
  expect(component.baseElement).toMatchSnapshot();
});

test('onSelect is fired on click', () => {
  const component = renderTabNavItem(initProps);
  fireEvent.click(component.getByText(/tab text/i));
  expect(onSelect).toHaveBeenCalledTimes(1);
  expect(component.baseElement).toMatchSnapshot();
});

