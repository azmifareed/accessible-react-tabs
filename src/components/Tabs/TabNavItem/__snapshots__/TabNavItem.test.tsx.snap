// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onSelect is fired on click 1`] = `
<body>
  <div>
    <button
      aria-controls="test-tab"
      aria-selected="false"
      class="sc-bdnxRM jvWqZJ"
      id="test"
      role="tab"
      tabindex="-1"
    >
      tab text
    </button>
  </div>
</body>
`;

exports[`renders TabNavItem correctly 1`] = `
<body>
  <div>
    <button
      aria-controls="test-tab"
      aria-selected="false"
      class="sc-bdnxRM jvWqZJ"
      id="test"
      role="tab"
      tabindex="-1"
    >
      tab text
    </button>
  </div>
</body>
`;

exports[`renders TabNavItem correctly when selected 1`] = `
<body>
  <div>
    <button
      aria-controls="test-tab"
      aria-selected="true"
      class="sc-bdnxRM dnvMwT"
      id="test"
      role="tab"
      tabindex="-1"
    >
      tab text
    </button>
  </div>
</body>
`;
